weave/tests/trace/test_table_query.py

```diff
@@ -0,0 +1,193 @@
+    result_digests = [r.digest for r in res.rows]
+
+    assert len(result_vals) == limit
+    assert result_digests == row_digests[:limit]
+    assert result_vals == [d for d in data[:limit]]
+
+
+def test_table_query_offset(client: WeaveClient):
+    digest, row_digests, data = generate_table_data(client, 10, 5)
+
+    offset = 3
+    res = client.server.table_query(
+        tsi.TableQueryReq(project_id=client._project_id(), digest=digest, offset=offset)
+    )
+
+    result_vals = [r.val for r in res.rows]
+    result_digests = [r.digest for r in res.rows]
+
+    assert len(result_vals) == len(data) - offset
+    assert result_digests == row_digests[offset:]
+    assert result_vals == [d for d in data[offset:]]
+
+
+def test_table_query_sort_by_column(client: WeaveClient):
+    digest, row_digests, data = generate_table_data(client, 10, 5)
+
+    res = client.server.table_query(
+        tsi.TableQueryReq(
+            project_id=client._project_id(),
+            digest=digest,
+            sort_by=[tsi.SortBy(field="id", direction="desc")],
+        )
+    )
+
+    result_vals = [r.val for r in res.rows]
+    result_digests = [r.digest for r in res.rows]
+
+    sorted_data = sorted(data, key=lambda x: x["id"], reverse=True)
+    assert result_vals == sorted_data
+    assert [r.val["id"] for r in res.rows] != [
+        d["id"] for d in data
+    ]  # Ensure order is different from original (assertion on the test itself)
+
+
+def test_table_query_sort_by_nested_column(client: WeaveClient):
+    digest, row_digests, data = generate_table_data(client, 10, 5)
+
+    res = client.server.table_query(
+        tsi.TableQueryReq(
+            project_id=client._project_id(),
+            digest=digest,
+            sort_by=[tsi.SortBy(field="nested_col.prop_a", direction="asc")],
+        )
+    )
+
+    result_vals = [r.val for r in res.rows]
+
+    sorted_data = sorted(data, key=lambda x: x["nested_col"]["prop_a"])
+    assert result_vals == sorted_data
+
+    assert [r.val["id"] for r in res.rows] != [
+        d["id"] for d in data
+    ]  # Ensure order is different from original (assertion on the test itself)
+
+
+def test_table_query_combined(client: WeaveClient):
+    digest, row_digests, data = generate_table_data(client, 20, 5)
+
+    limit = 5
+    offset = 2
+    res = client.server.table_query(
+        tsi.TableQueryReq(
+            project_id=client._project_id(),
+            digest=digest,
+            limit=limit,
+            offset=offset,
+            sort_by=[tsi.SortBy(field="id", direction="desc")],
+        )
+    )
+    result_vals = [r.val for r in res.rows]
+    result_digests = [r.digest for r in res.rows]
+
+    sorted_data = sorted(data, key=lambda x: x["id"], reverse=True)
+    expected_data = sorted_data[offset : offset + limit]
+
+    assert len(res.rows) == limit
+    assert result_vals == expected_data
+
+
+def test_table_query_multiple_sort_criteria(client: WeaveClient):
```

Comment:
might also want to add a test for paging/offset with a query that has a non-existent digest/table.